### PR TITLE
[ironnet-262] Update to IronDefense integration. Adding get events and alerts

### PR DESCRIPTION
**Pull request by PaulMcHugh**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Update IronDefense integration to be able to retrieve events and alerts.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [x] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

## Demisto Partner?
- [x] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by ronykoz** [comment]
@PaulMcHugh Thanks for the contribution, i see that you've checked the `In Progress` checkbox, please leet me know when it is ready for a review and i will take a look at your contribution.

**Comment by PaulMcHugh** [comment]
@ronykoz My mistake, it's ready for review now.

**Comment by ronykoz** [comment]
@PaulMcHugh i just saw that you've merged from master, it added some files to your PR, but no worries, i will try to fix that by myself later today or tomorrow, please ignore them for now

**Comment by PaulMcHugh** [comment]
@ronykoz Thank you, I was trying to fix the bad object error I was getting in the build check. I assumed I need to rebase off master.

File: Integrations/IronDefense/IronDefense.yml [hunk 1]
```diff
@@ -21,10 +21,102 @@ configuration:
...
+  - Recon
+  - Other
+  required: false
+  type: 16
+- display: Alert Notification SubCategories to Exclude (enter in CSV format).
+  name: alertSubCategories
+  options:
+  - ''
+  required: false
+  type: 12
```
[the remark this answers is not in the document]
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.yml]: We don't want options for sub categories. Removing this.

File: Integrations/IronDefense/IronDefense.yml [hunk 2]
```diff
@@ -21,10 +21,102 @@ configuration:
...
+  - Access
+  - Recon
+  - Other
+  required: false
+  type: 16
+- display: Event Notification SubCategories to Exclude (enter in CSV format).
+  name: eventSubCategories
+  options:
+  - ''
+  required: false
```
[the remark this answers is not in the document]
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.yml]: Removing this.

File: Integrations/IronDefense/IronDefense.yml [hunk 3]
```diff
@@ -186,11 +278,226 @@ script:
...
+      name: event_id
+      required: true
+      secret: false
+    deprecated: false
+    description: Retrieves an IronDefense event.
+    execution: false
+    name: irondefense-get-event
+    outputs:
+    - contextPath: event
+      description: The IronDefense event.
```
[the remark this answers is not in the document]
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.yml]: Okay, I'm filling in the context path I'll update those in the next commit

**Comment by PaulMcHugh** [comment]
@ronykoz I've made the updates you suggested. 
Would you like me to rebase this branch back to the other release to remove the additional commits on here.
And you mention you'd like to see the IronDefense integration in action, we are currently working on updating our demo video and can forward that along to you if you'd like.

**Comment by ronykoz** [comment]
@PaulMcHugh is it review ready?

**Comment by PaulMcHugh** [comment]
@ronykoz Yes this is ready for review. Just I see the circleci is failing because the demisto-sdk branch is gone.

File: Integrations/IronDefense/IronDefense.yml [hunk 4]
```diff
@@ -182,15 +310,679 @@ script:
...
+    - contextPath: IronDefense.Event.context.name
+      description: The name of the context table.
+      type: String
+    - contextPath: IronDefense.Event.context.columns
+      description: Column values for the context table.
+      type: String
+  - arguments:
+    - default: false
+      description: The list of Alert IDs to filter by.
+      isArray: false
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.yml:431]: if it's a list than this one should be true

File: Integrations/IronDefense/IronDefense.yml [hunk 5]
```diff
@@ -182,15 +310,679 @@ script:
...
+      name: alert_id
+      required: false
+      secret: false
+    - default: false
+      description: The list of Alert categories to filter by.
+      isArray: false
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.yml:437]: same here, and everywhere else in the yml

File: Integrations/IronDefense/IronDefense.yml [hunk 6]
```diff
@@ -182,15 +310,679 @@ script:
...
+  longRunning: false
+  longRunningPort: false
   runonce: false
   script: '-'
-  type: python
   subtype: python3
+  type: python
 tests:
-- IronDefenseTest
+- No tests
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.yml:988]: why did you change this

File: Integrations/IronDefense/IronDefense.py [hunk 7]
```diff
@@ -1,10 +1,15 @@
+import functools
+from typing import List, Dict
 import demistomock as demisto
-from CommonServerPython import *
-from CommonServerUserPython import *
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.py:3]: no need to remove those lines, we are handling them in the artifact generation

File: Integrations/IronDefense/IronDefense.py [hunk 8]
```diff
@@ -241,14 +403,234 @@ def report_observed_bad_activity(self, name, description='', ip='', domain='',
...
+        LOGGER.debug("Ingestion of all notifications (Dome, Alert, Event) is disabled, not fetching")
+    else:
+        if disable_dome_notifs:
+            LOGGER.debug('Ingestion of Dome Notifications is disabled')
+        else:
+            incs = IRON_DEFENSE.fetch_dome_incidents(dome_categories, dome_limit)
+            incidents.extend(incs)
+            # If the limit was reached, poll again
+            poll_count = 1
+            while len(incs) == dome_limit and poll_count < 10:
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.py:600]:
```suggestion
            while len(incs) <= dome_limit and poll_count < 10:
```

File: Integrations/IronDefense/IronDefense.py [hunk 9]
```diff
@@ -241,14 +403,234 @@ def report_observed_bad_activity(self, name, description='', ip='', domain='',
...
+
+        if disable_alert_notifs:
+            LOGGER.debug('Ingestion of Alert Notifications is disabled')
+        else:
+            incs = IRON_DEFENSE.fetch_alert_incidents(alert_categories, alert_subcategories, alert_severity_lower,
+                                                      alert_severity_upper, alert_limit, alert_actions)
+            incidents.extend(incs)
+            # If the limit was reached, poll again
+            poll_count = 1
+            while len(incs) == alert_limit and poll_count < 10:
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.py:613]:
```suggestion
            while len(incs) <= alert_limit and poll_count < 10:
```

File: Integrations/IronDefense/IronDefense.py [hunk 10]
```diff
@@ -241,14 +403,234 @@ def report_observed_bad_activity(self, name, description='', ip='', domain='',
...
+
+        if disable_event_notifs:
+            LOGGER.debug('Ingestion of Event Notifications is disabled')
+        else:
+            incs = IRON_DEFENSE.fetch_event_incidents(event_categories, event_subcategories, event_severity_lower,
+                                                      event_severity_upper, event_limit, event_actions)
+            incidents.extend(incs)
+            # If the limit was reached, poll again
+            poll_count = 1
+            while len(incs) == event_limit and poll_count < 10:
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.py:627]:
```suggestion
            while len(incs) <= event_limit and poll_count < 10:
```

File: Integrations/IronDefense/IronDefense.py [hunk 11]
```diff
@@ -301,14 +684,272 @@ def report_observed_bad_activity_command():
...
+                       outputs={
+                           'IronDefense.Event(val.id == obj.id)': event,
+                       },
+                       raw_response=event)
+
+    # Send constraints
+    constraint = results.get('constraint')
+    return_outputs(readable_output=tableToMarkdown('Query Constraints', constraint),
+                   outputs={
+                   'IronDefense.Query.GetEvents': constraint,
```
Comment (ronykoz) [inline Integrations/IronDefense/IronDefense.py:747]: the constraints should be under the GetEvents in the context?

File: Integrations/IronDefense/IronDefense.yml (hunk 4 above)
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.yml:431]: For these we're not taking it as an array. We're  taking a CSV string and splitting it.

File: Integrations/IronDefense/IronDefense.yml (hunk 6 above)
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.yml:988]: I think this just got lost. I'll revert this line

File: Integrations/IronDefense/IronDefense.py (hunk 8 above)
Comment (PaulMcHugh) [reply Integrations/IronDefense/IronDefense.py:600]: I don't agree with this change. We only want to poll again if the previous call reached the limit. If the last call came in under the limit we assume we're at the end of the queue.
